tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,344 @@
+
+    mutable_input_post_processing = '\n'.join([
+        f"""
+      {a.name}.replace_(tmp_output);
+      at::functionalization::impl::commit_update({a.name});"""
+        for a in f.func.arguments.flat_non_out
+        if a.annotation and a.annotation.is_write and a.type.is_tensor_like()])
+
+    return f"""
+      {sync_tensor_args}
+      {unwrap_tensor_args_str(f)}
+      {return_type} tmp_output;
+      {{
+          at::AutoDispatchSkipFunctionalize guard;
+          // The functionalization pass explicitly doesn't pass out= parameters to the redispatch
+          tmp_output = at::redispatch::{functional_api_name}(
+            {', '.join([keyset] + [arg_name(a) for a in f.func.arguments.flat_non_out])});
+      }}
+      {mutable_input_post_processing}
+      {maybe_return}{return_names_str(f)};"""
+
+
+
+def emit_registration(f: NativeFunction, native_api_name: Optional[str]) -> str:
+    if f.has_composite_implicit_autograd_kernel:
+        assert native_api_name is not None, f"Found a view op ({f.func.name}) with no cpu or composite kernel." \
+            f" The functionalization pass codegen needs to know the at::native::{f.func.name} kernel name of every view op," \
+            f" which is currently does by looking up the cpu/composite kernel kernel name of the op. If you encountered a view op" \
+            f" without a cpu/composite kernel, you'll need to change this logic"
+        sig = DispatcherSignature.from_schema(f.func)
+        # Note [Composite view ops in the functionalization pass]
+        # We don't need to worry about implemententing functionalization kernels for views with
+        # CompositeImplicitAutograd kernels, because we can just decompose them into their base operators.
+        # We can't just opt the entire Functionalization dispatch key into the composite keyset though,
+        # because we don't want to decompose non-view ops that are composite, like `at::ones`.
+        registration_str = f'static_cast<{sig.ptr_type()}>(at::native::{native_api_name})'
+    else :
+        registration_str = f'TORCH_FN(functionalization::{type_wrapper_name(f)})'
+
+    return f"""
+m.impl("{f.func.name}",
+       {registration_str}
+);
+"""
+
+def emit_definition(f: NativeFunction, g: Optional[NativeFunctionsGroup], native_api_name: Optional[str]) -> str:
+    # order is important here, ops that are both views and mutations should hit the view path.
+    if is_view_op(f):
+        assert native_api_name is not None, f"Found a view op ({f.func.name}) with no cpu or composite kernel." \
+            f" The functionalization pass codegen needs to know the at::native::{f.func.name} kernel name of every view op," \
+            f" which is currently does by looking up the cpu/composite kernel kernel name of the op. If you encountered a view op" \
+            f" without a cpu/composite kernel, you'll need to change this logic"
+        body_str = emit_view_functionalization_body(f, g, native_api_name)
```

Comment:
Although it involves plumbing the backend indexes to this site (probably most easily done by making these methods on the class), it seems better to only compute `native_api_name` when you have a view op, and then immediately assert when you are trying to pull it out (so you can, e.g., report what kernels are actually in the dispatch table). The assert here is missing all the important information to give a good error message here; even worse, the error message says you need to "change logic", but the logic is no where near this assert, it's elsewhere...